internal/metrics/metrics.go

```diff
@@ -127,3 +129,48 @@ func (s *ScheduleTimer) Mark(ctx context.Context, role, outcome string) {
 	metrics.FromContext(ctx).Histogram(
 		schedulingRetryTime, sparkRoleTag, outcomeTag, s.instanceGroupTag, s.retryTag).Update(now.Sub(s.lastSeenTime).Nanoseconds())
 }
+
+// ReportCrossZoneMetric reports metric about cross AZ traffic between pods of a spark application
+func ReportCrossZoneMetric(ctx context.Context, driverNodeName string, executorNodeNames []string, nodes []*v1.Node) {
+	executorNodesSet := make(map[string]interface{})
+	for _, n := range executorNodeNames {
+		executorNodesSet[n] = nil
+	}
+
+	zonesCounter := make(map[string]int)
+	for _, n := range nodes {
+		if _, ok := executorNodesSet[n.Name]; ok {
+			executorZone, ok := n.Labels[nodeZoneLabel]
+			if !ok {
+				svc1log.FromContext(ctx).Warn("zone label not found for node", svc1log.SafeParam("nodeName", n.Name))
+				return
+			}
+			zonesCounter[executorZone]++
+		} else if n.Name == driverNodeName {
+			driverZone, ok := n.Labels[nodeZoneLabel]
+			if !ok {
+				svc1log.FromContext(ctx).Warn("zone label not found for node", svc1log.SafeParam("nodeName", n.Name))
+				return
+			}
+			zonesCounter[driverZone]++
+		}
+	}
+
+	metrics.FromContext(ctx).Histogram(crossAzTraffic).Update(int64(crossZoneTraffic(zonesCounter)))
+}
+
+// crossZoneTraffic calculates the total number of pairs of pods, where the 2 pods are in different zones.
+// A pair represents potential cross-zone traffic, which we want to avoid.
+func crossZoneTraffic(zonesCounter map[string]int) int {
```

Comment:
I wonder if we just do math here: It would be `C(total, 2) - C(num_az1, 2) - C(num_az2, 2)` for a two az cluster

Reply:
I don't think that would be  more efficient though, and given you explain what this does above, I think this is fine to keep